lib/api/confighandler.go

```diff
@@ -0,0 +1,375 @@
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+const configBase = "/rest/config/"
+
+type configHandler struct {
+	*httprouter.Router
+	id  protocol.DeviceID
+	cfg config.Wrapper
+	mut sync.Mutex
+}
+
+func newConfigHandler(id protocol.DeviceID, cfg config.Wrapper) http.Handler {
+	c := &configHandler{
+		Router: httprouter.New(),
+		id:     id,
+		cfg:    cfg,
+		mut:    sync.NewMutex(),
+	}
+
+	c.registerConfig(configBase)
+	c.registerConfigInsync(configBase + "insync")
+	c.registerFolders(configBase + "folders")
+	c.registerDevices(configBase + "devices")
+	c.registerFolder(configBase + "folders/:id")
+	c.registerDevice(configBase + "devices/:id")
+	c.registerOptions(configBase + "options")
+	c.registerLDAP(configBase + "ldap")
+	c.registerGUI(configBase + "gui")
+
+	return c
+}
+
+func (c *configHandler) registerConfig(path string) {
+	legacyPath := "/rest/system/config"
+
+	handler := func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.RawCopy())
+	}
+	c.GET(path, handler)
+	c.GET(legacyPath, handler)
+
+	handler = func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustConfig(w, r)
+	}
+	c.POST(path, handler)
+	c.PUT(path, handler)
+	c.POST(legacyPath, handler)
+	c.PUT(legacyPath, handler)
+}
+
+func (c *configHandler) registerConfigInsync(path string) {
+	legacyPath := "/rest/system/config/insync"
+	handler := func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, map[string]bool{"configInSync": !c.cfg.RequiresRestart()})
+	}
+	c.GET(path, handler)
+	c.GET(legacyPath, handler)
+}
+
+func (c *configHandler) registerFolders(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.FolderList())
+	})
+	handler := func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		var waiter config.Waiter
+		folders := make([]config.FolderConfiguration, 0)
+		err := unmarshalTo(r.Body, &folders)
+		if err == nil {
+			waiter, err = c.cfg.SetFolders(folders)
+		}
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		c.finish(w, waiter)
+	}
+	c.POST(path, handler)
+	c.PUT(path, handler)
+}
+
+func (c *configHandler) registerDevices(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.DeviceList())
+	})
+	handler := func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		var waiter config.Waiter
+		devices := make([]config.DeviceConfiguration, 0)
+		err := unmarshalTo(r.Body, &devices)
+		if err == nil {
+			waiter, err = c.cfg.SetDevices(devices)
+		}
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		c.finish(w, waiter)
+	}
+	c.POST(path, handler)
+	c.PUT(path, handler)
+}
```

Comment:
According for rest-fu, PUT/POST should be used for singular resources.

I guess following the rest logic for PUT and POST I would interpret this as:
PUT - Update the folder list to look like this (adding/removing/modifying folders)
POST - Create the folders in the array?


I think this is somewhat gray ground, so I am not sure if we want to support PUT/POST for collection endpoints all together, but PUT and POST doing the same is definately not cool.

Reply:
I used both because from what I read `PUT` is right for our use-cases (e.g. all the operations are idempotent), but we still did use `POST` on the previous config endpoint. However I'll just keep `POST` on the anyway legacy `/rest/system/config` and use `PUT` everywhere else exclusively.

Reply:
I think this is old and resolved, but my understanding of things is that if `/config/devices/` is a collection, POST to `/config/devices/` could add a new device, while PUT to `/config/devices/$someDeviceID` would change or add the device with that specific ID. PUT to `/config/devices/` to replace the entire collection of devices seems kosher.